### ISSUE TITLE
[red-knot] Return type checking in presence of calls to `Never`/`NoReturn`

### ISSUE THREAD

**Issue by sharkdp** [astral-sh/ruff#16700]
### Summary

The following code emits a `invalid-return-type` diagnostic, but shouldn't:
```py
from typing import NoReturn
import sys

def f() -> NoReturn:
#          ^^^^^^^^ Function can implicitly return `None`, which is not assignable to return type `Never`
    sys.exit(1)
```

`sys.exit` is annotated with returning `NoReturn`, but the problem is that we don't have an explicit `return` statement here.

This is related to unreachable code analysis. We already have some special handling for detecting whether or not the end of scope is reachable (i.e. replacing `sys.exit(1)` with `while True: pass` works fine), but this does not take calls to functions returning `Never`/`NoReturn` into account.

**Comment by abhijeetbodas2001** [comment]
Hi David, I would like on this. Would this be an OK first issue to pick up?

**Comment by sharkdp** [comment]
You are certainly welcome to try! However, I do not think that this is a particularly beginner friendly task. We might have some issues labeled with "help wanted" that might be more suitable for a first contribution.

**Comment by abhijeetbodas2001** [comment]
Got it. I will try to find another issue, or another minor task in this area (I do see some TODOs in the tests, for example).

However, one thing is not clear to me. You mention in the issue description that replacing `sys.exit(1)` with `while True: pass` does not give an error. I tried that out, and indeed there is no error.

However, the following **also** does not give any error:

```py
def f() -> int:
  while True:
    pass
```

(Notice that I changed the return value to `int`.)

Here, if the "`while True` never returns anything" inference was being done correctly, the above should have errored out, right? 

Moreover, the following gives an error:

```py
# "Object of type `Literal["foo"]` is not assignable to return type `int`"
def f() -> int:
  while True:
    pass

  return "foo"
```

... which makes me wonder if `red-knot` does not correctly understand the `while True: pass` case at all. Am I missing something?

**Comment by sharkdp** [comment]
@abhijeetbodas2001 Sorry for responding so late, I must have missed the notification about your original message while I was out on vacation.

> Here, if the "`while True` never returns anything" inference was being done correctly, the above should have errored out, right?

I don't think so. A function annotated with `int` that loops endlessly is not incorrect. You can think of the endless loop as returning `Never`, but `Never` is assignable to the return type `int` (`Never` is assignable to any type), so nothing wrong there.

> Moreover, the following gives an error:

Yes, we don't silence *all* errors in unreachable code. And in this case, it's arguable if that would be an improvement. If the loop were not endless, that return statement *would* be wrong. So flagging it does not seem like unwanted behavior to me.

**Comment by abhijeetbodas2001** [comment]
> (Never is assignable to any type)

Ah! I missed this. That makes sense then. Thanks!
